src/outside/grammar_transformation.cc

```diff
@@ -0,0 +1,969 @@
+        for (std::list<Alt::Base*>::iterator child =
+             v_block.block->alts.begin();
+             child != v_block.block->alts.end(); ++child) {
+          Alt::Base *clone = (*child)->clone();
+
+          // since parent is lhs non-terminal and block itself will be removed,
+          // children will become top level alternatives
+          clone->top_level = Bool(true);
+
+           // don't forget to carry over filters ...
+          clone->filters.insert(clone->filters.end(),
+              v_block.block->filters.begin(),
+              v_block.block->filters.end());
+
+          // ... and filters for multitrack
+          clone->multi_filter.insert(clone->multi_filter.end(),
+              v_block.block->multi_filter.begin(),
+              v_block.block->multi_filter.end());
+
+          // insert new (partially, since it can still hold further Blocks)
+          // alternative into rhs of the NT
+          nt->alts.insert(topalt, clone);
+        }
+
+        nt->alts.remove(*topalt);
+      } else {
+        throw LogError("each Alt::Block should have a parent!");
+      }
+
+      // check if there exist further Alt::Blocks, if not, we exit the while
+      // loop
+      v_block = FindFirstBlock();
+      nt->traverse(v_block);
+    }
+  }
+}
+
+/* Counts the number of components a rule will inject into parse-trees, e.g.
+ *   hairpin = hl(BASE, REGION, BASE) : increase by 1 due to hl
+ *   plus = CHAR('+') : no increase since only terminal used on rhs
+ *   weak = hairpin | bulgeL : increase by 2 due to 2 non-terminals
+ * At least one component is necessary to subject the production rule to
+ * outside generation and/or outside axiom determination
+ */
+struct Count_parseComponents : public Visitor {
+  unsigned int number_components = 0;
+
+  Count_parseComponents() {
+    number_components = 0;
+  }
+
+  void visit_begin(Alt::Simple &alt) {
+    if (!alt.is_terminal()) {
+      number_components++;
+    }
+  }
+  void visit(Alt::Link &alt) {
+    if (alt.nt->is(Symbol::NONTERMINAL)) {
+      number_components++;
+    }
+  }
+};
+
+/* let all Alt::Simple grammar elements know which left hand NT called this
+ * alternative. Necessary to construct proper outside guards, which need to know
+ * lhs NT table dimensions.
+ */
+struct Propagate_lhsNT : public Visitor {
+  Symbol::NT *lhsNT;
+
+  explicit Propagate_lhsNT(Symbol::NT *lhsNT) : lhsNT(lhsNT) {
+  }
+
+  void visit_begin(Alt::Simple &alt) {
+    alt.outside_lhsNT = this->lhsNT;
+  }
+};
+
+
+/* iterates through the rhs alternatives of an NT and creates a clone of an
+ * alternative where ONE (but not all) rhs NT is swapped with the lhs NT, e.g.
+ *   struct = cadd(dangle, weak) | sadd(BASE, struct) will result in
+ * a) outside_dangle = cadd(outside_struct, weak)
+ * b) outside_weak = cadd(dangle, outside_struct)
+ * c) outside_struct = sadd(BASE, outside_struct) */
+struct Flip_lhs_rhs_nonterminals : public Visitor {
```

Comment:
The objects that get allocated in this class can be deleted when an object of this class gets destroyed:
```
~Flip_lhs_rhs_nonterminals() {
    delete alt_clones;
    delete lhs_nt->name;
    delete lhs_nt;
  }
```

Reply:
why do we need to delete `lhs_nt->name` if we delete `lhs_nt` in the line below?

Reply:
Because the `NT` object doesn't free the memory of the `name` ptr (or any other ptr) it stores when it gets destructed. So after destruction of the `NT`, only the ptr to `name` will be deleted, but no the actual data `name` pointed to.

Reply:
Looks like trying to free the memory in the destructors doesn't work as intended. I scanned the code a bit more and think that the data that the ptrs of this object point to is used after the object has already been destroyed (e.g. `convert_to_outside` in `grammar_transformation.cc`, lines 886ff.), which can obviously lead to problems if we've deleted that data already. I guess we will have to live with not freeing that memory here.